src/object/cli_obj.c

```diff
@@ -1005,207 +1050,117 @@ obj_retry_cb(tse_task_t *task, struct dc_object *obj,
 	return rc;
 }
 
-/* shard update/punch auxiliary args, must be the first field of
- * shard_update_args and shard_punch_args.
- */
-struct shard_auxi_args {
-	struct dc_object	*obj;
-	struct obj_auxi_args	*obj_auxi;
-	uint32_t		 shard;
-	uint32_t		 target;
-	uint32_t		 map_ver;
-	uint32_t		 start_shard;
-};
-
 struct obj_list_arg {
 	struct dc_object	*obj;
 	daos_anchor_t		*anchor;	/* anchor for record */
 	daos_anchor_t		*dkey_anchor;	/* anchor for dkey */
 	daos_anchor_t		*akey_anchor;	/* anchor for akey */
 };
 
-struct shard_update_args {
-	struct shard_auxi_args	 auxi;
-	daos_epoch_t		 epoch;
-	struct dtx_id		 dti;
-	daos_key_t		*dkey;
-	uint64_t		 dkey_hash;
-	unsigned int		 nr;
-	daos_iod_t		*iods;
-	d_sg_list_t		*sgls;
-};
-
+/* prepare the bulk handle(s) for obj request */
 static int
-shard_result_process(tse_task_t *task, void *arg)
+obj_bulk_prep(d_sg_list_t *sgls, unsigned int nr, bool bulk_bind,
+	      crt_bulk_perm_t bulk_perm, tse_task_t *task,
+	      struct obj_auxi_args *obj_auxi)
 {
-	struct obj_auxi_args	*obj_auxi = (struct obj_auxi_args *)arg;
-	struct shard_auxi_args	*shard_auxi;
-	int			 ret = task->dt_result;
+	crt_bulk_t	*bulks;
+	int		 i = 0;
+	int		 rc = 0;
 
-	/*
-	 * Check shard IO task's completion status:
-	 * 1) if succeed just stores the highest replied pm version.
-	 * 2) if any shard got retryable error, mark the IO as retryable.
-	 * 3) for the un-retryable failure, store it in obj_auxi->result.
-	 */
-	shard_auxi = tse_task_buf_embedded(task, sizeof(*shard_auxi));
-	if (ret == 0) {
-		if (obj_auxi->map_ver_reply < shard_auxi->map_ver)
-			obj_auxi->map_ver_reply = shard_auxi->map_ver;
-	} else if (obj_retry_error(ret)) {
-		D_DEBUG(DB_IO, "shard %d ret %d.\n", shard_auxi->shard, ret);
-		obj_auxi->io_retry = 1;
-		if (ret == -DER_INPROGRESS)
-			obj_auxi->retry_with_leader = 1;
-	} else {
-		/* for un-retryable failure, set the err to whole obj IO */
-		D_DEBUG(DB_IO, "shard %d ret %d.\n", shard_auxi->shard, ret);
-		obj_auxi->result = ret;
+	D_ASSERTF(nr >= 1, "invalid nr %d.\n", nr);
+	D_ALLOC_ARRAY(bulks, nr);
+	if (bulks == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	/* create bulk handles for sgls */
+	for (; i < nr; i++) {
```

Comment:
you could move sgls != NULL to the loop above
```for (; sgls != NULL && i < nr; i++)```

Reply:
agreed, can change it when refresh it next time or at next patch.